infrastructure/equinix-metal/main.tf

```diff
@@ -0,0 +1,61 @@
+terraform {
+  required_providers {
+    equinix = {
+      source  = "equinix/equinix"
+      version = "1.13.0"
+    }
+  }
+
+  backend "s3" {
+    bucket  = "green-reviews-state-bucket"
+    key     = "opentofu/terraform.tfstate"
+    region  = "eu-central-1"
+    encrypt = true
+  }
+}
+
+provider "equinix" {
+  auth_token = var.equinix_auth_token
+}
+
+resource "equinix_metal_project_ssh_key" "ssh_key" {
+  name       = var.cluster_name
+  project_id = var.project_id
+  public_key = var.ssh_public_key
+}
+
+resource "equinix_metal_device" "control_plane" {
+  hostname            = "${var.cluster_name}-control-plane"
+  plan                = var.device_plan
+  metro               = var.device_metro
+  operating_system    = var.device_os
+  billing_cycle       = var.billing_cycle
+  project_id          = var.project_id
+  depends_on          = [equinix_metal_project_ssh_key.ssh_key]
+  project_ssh_key_ids = [equinix_metal_project_ssh_key.ssh_key.id]
+}
+
+resource "local_file" "inventory" {
+  filename = "${path.module}/inventory.ini"
+
+  content = <<-EOF
+    [control_plane]
+    ${equinix_metal_device.control_plane.hostname} ansible_host=${equinix_metal_device.control_plane.access_public_ipv4} ansible_user=root
+  EOF
+
+  depends_on = [
+    equinix_metal_device.control_plane
+  ]
+}
+
+resource "null_resource" "ansible_playbook" {
+  triggers = {
+    always_run = "${timestamp()}"
```

Comment:
Right now the Ansible playbook is run each time. Do you think this is OK? Or can we detect if the playbook has changed?

Reply:
- If then infrastructure changes then configuration needs to be applied
- If no change in infrastructure and we are applying will it affect existing cluster?

May be this conditional statement can help
```
If (infrastructure changes)
  Apply scripts
Else 
  If (scripts changed)
    Apply scripts
```

Reply:
Not really sure, never used null_resources
but from few projects they use `null_resources` for executing some commands
and for user_data its sepecific to VM specific first time boot scripts

Reply:
need to check it out

Reply:
refering: https://spacelift.io/blog/terraform-provisioners#how-to-use-terraform-provisioners

we can use the remote exec to get the k3s token
and then we can use the out put from this to insert in another resource for (VM for workernodes) which have to wait for both the controlplane node with k3s installed and the k3s token

Reply:
few projects they use null_resources for executing some commands
and for user_data its sepecific to VM specific first time boot scripts

Reply:
Also we can refer to this https://docs.k3s.io/installation/airgap#installing-k3s-in-an-air-gapped-environment

we can do is create a random string using terraform resource and use that as the k3s token

by this we eliminate the use of custom commands to be executed and we can use that token to join other workernodes

Reply:
and for installing cilium we can just pass the flag for nocni and then we can install cilium using helm install

Reply:
> few projects they use null_resources for executing some commands
> and for user_data its specific to VM specific first time boot scripts

Yes, I used it for triggering Ansible but I wasn't sure how widely used it is and better to use user_data. I'll remove.

> we can do is create a random string using terraform resource and use that as the k3s token

Nice! I like the idea of generating the k3s token.

Reply:
Can I open a PR for `rossf7:wip-opentofu-spike` add some scripts there for k3s?
or we should wait for everyone opnions?

Reply:
@dipankardas011 Thanks if you can open a PR with the scripts I think that would be helpful. So we can see what it would look like using k3s.

We do need to get more opinions and discuss the tools at the next WG meeting. I've added a note to the PR description that the tools used are provisional and still to be agreed.

Reply:
okay np I will wait for more opnions then I will create PR for the scripts
to avoid dual work 😄

Reply:
Good call 👍 better to wait till we get more feedback

Reply:
till now Here is what it looks like
```tf
resource "random_string" {
   // some value
}

resource "controlplanevm" {

  user_data = << EOF
  #!/bin/bash
  curl -sfL https://get.k3s.io | INSTALL_K3S_CHANNEL="1.28-..." K3S_TOKEN=${random_string.output} sh -s - server \
	--node-taint CriticalAddonsOnly=true:NoExecute \
	--flannel-backend=none \
	--disable-network-policy \
	--tls-san %s
  EOF
}

resource "WorkerNodesvm" {
  for loop for 1 .. N

  depends_on = [ controlplanevm ]

  user_data = << EOF
  #!/bin/bash
  curl -sfL https://get.k3s.io | INSTALL_K3S_CHANNEL="1.28-..." sh -s - agent --token ${random_string.output} --server https://${controlplanevm.networks.ip}:6443
  EOF
}
```

Reply:
one issue is how to know if the controlplane is configured before we execute the worknodes creation step
one solution which come to my mind is use another resource which checks if k3s systemd service is op and running

Reply:
@dipankardas011 that looks good and a null resource to check k3s is up on the control plane node makes sense to me.

I just pushed some small changes to remove Ansible and create worker nodes based on a list of node names.

Reply:
nice